charms/candid/src/charm.py

```diff
@@ -113,28 +126,43 @@ def _start(self, event: StartEvent):
         if self.snap_running:
             self.set_status_and_log("Ready", ActiveStatus)
 
-    def _config_changed(self, event: StartEvent):
+    def _on_upgrade_charm(self, event):
+        self._install(event)
+        self._start(event)
+
+    def _config_changed(self, event):
         """
         Updates snap internal configuration.
         """
         self._update_config_and_restart(event)
 
+    @requires_state
     def _update_config_and_restart(self, event):
         if not self._check_config(event):
             return
 
         config_values = {
             key: value for key, value in self.config.items() if value
         }
-        if self._stored.db_uri:
+        if self._state.db_uri:
             config_values["storage"] = {
                 "type": "postgres",
-                "connection-string": self._stored.db_uri,
+                "connection-string": self._state.db_uri,
             }
         else:
-            logging.info("db_uri not stored")
+            logging.info("missing postgres relation")
```

Comment:
We should enter a blocked state here and return

Reply:
Though granted lower down if the DB connection string is removed we call this function and if we were to return early here we wouldn't stop the Candid snap so it's a bit tricky.

Reply:
removed this check as it is not needed.. code calls self._check_config(event) which will return False if postgres connection string is missing and go into Waiting status